app/utils/sql/SqlInterpolation.scala

```diff
@@ -0,0 +1,248 @@
+
+        protected def setParam(st: PreparedStatement): Unit = {
+          val pp = new PositionedParameters(st)
+          values.foreach(_.setParameter(pp))
+        }
+
+        protected def extractValue(rs: PositionedResult): R = resultConverter(rs)
+      }
+
+      override def getDumpInfo = DumpInfo(DumpInfo.simpleNameFor(getClass), mainInfo = s"[$debugInfo]")
+
+      protected[this] def createBuilder: mutable.Builder[R, Vector[R]] = Vector.newBuilder[R]
+    }
+
+  def asUpdate: SqlAction[Int, NoStream, Effect] = as[Int](GetUpdateValue).head
+}
+
+object SqlToken {
+  def join(values: List[Either[SqlValue, SqlToken]], sep: String): SqlToken = {
+    val outputSql = mutable.StringBuilder.newBuilder
+    val outputValues = ListBuffer[SqlValue]()
+    for (i <- values.indices) {
+      val value = values(i)
+      value match {
+        case Left(x) =>
+          outputSql ++= x.placeholder
+          outputValues += x
+        case Right(x) =>
+          outputSql ++= x.sql
+          outputValues ++= x.values
+      }
+      if (i < values.length - 1) {
+        outputSql ++= sep
+      }
+    }
+    SqlToken(sql = outputSql.toString, values = outputValues.toList)
+  }
+
+  def tuple(values: Seq[Any]): SqlToken = {
+    val sqlValues = values.map(SqlValue.makeSqlValue)
+    SqlToken(sql = s"(${sqlValues.map(_.placeholder).mkString(", ")})", values = sqlValues.toList)
+  }
+
+  def tupleList(values: Seq[Seq[Any]]): SqlToken = {
+    val sqlValueLists = values.map(list => list.map(SqlValue.makeSqlValue))
+    SqlToken(sql = sqlValueLists.map(list => s"(${list.map(_.placeholder).mkString(", ")})").mkString(", "),
+             values = sqlValueLists.flatten.toList)
+  }
+
+  def raw(s: String): SqlToken = SqlToken(s)
+
+  def empty: SqlToken = raw("")
+
+  def identifier(id: String): SqlToken = raw('"' + id + '"')
+}
+
+trait SqlValue {
+  def setParameter(pp: PositionedParameters): Unit
+
+  def placeholder: String = "?"
+
+  def debugInfo: String
+}
+
+object SqlValue {
+
+  @tailrec
+  def makeSqlValue(p: Any): SqlValue =
+    p match {
+      case x: SqlValue => x
+      case x: String   => StringValue(x)
+      case x: Option[_] =>
+        x match {
+          case Some(y) => makeSqlValue(y)
+          case None    => NoneValue()
+        }
+      case x: Short          => ShortValue(x)
+      case x: Int            => IntValue(x)
+      case x: Long           => LongValue(x)
+      case x: Float          => FloatValue(x)
+      case x: Double         => DoubleValue(x)
+      case x: Boolean        => BooleanValue(x)
+      case x: Instant        => InstantValue(x)
+      case x: FiniteDuration => DurationValue(x)
+      case x: ObjectId       => ObjectIdValue(x)
+      case x: JsValue        => JsonValue(x)
+    }
+}
+
+case class StringValue(v: String) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setString(v)
+
+  override def debugInfo: String = "'" + v + "'"
```

Comment:
the comment says that debugInfo should be pastable in an sql editor – I guess this might mean that v should be escaped here?